wp/lib/bap_wp/src/precondition.ml

```diff
@@ -1388,3 +1396,192 @@ let construct_pointer_constraint (l_orig : Constr.z3_expr list) (env1 : Env.t)
+  Some pre
+
+(* Gets the body of a loop without the header node containing the loop
+   condition. *)
+let get_body (graph : Graphs.Ir.t) (loop : Graphs.Ir.Node.t group) (blk : Blk.t)
+  : Graphs.Ir.t =
+  let outside_loop node =
+    let label = Graphs.Ir.Node.label node in
+    (Blk.equal label blk) || (not @@ in_loop loop (Term.tid label))
+  in
+  (* Removes the block that represents the loop header which determines
+     whether to jump to the loop or exit. *)
+  let enter_node _ node g =
+    if outside_loop node then
+      Graphs.Ir.Node.remove node g
+    else
+      g
+  in
+  (* Removes edges that exit the loop. *)
+  let enter_edge _ edge g =
+    let dst = Graphs.Ir.Edge.dst edge in
+    if outside_loop dst then
+      Graphs.Ir.Edge.remove edge g
+    else
+      g
+  in
+  Graphlib.depth_first_search (module Graphs.Ir)
+    ~enter_node ~enter_edge ~init:graph graph
+
+(* Splits the loop into the condition and the body of the loop. Assumes there
+   are no breaks in the loop. *)
+let split_loop (loop : Graphs.Ir.Node.t group) (graph : Graphs.Ir.t) (env : Env.t)
+  : (Constr.t * Sub.t) option =
+  let sub = Sub.of_cfg graph in
+  Seq.find_map (Term.enum blk_t sub) ~f:(fun blk ->
+      match get_cond loop blk env with
+      | Some cond ->
+        let body = Sub.of_cfg (get_body graph loop blk) in
+        Some (cond, body)
+      | None -> None)
+
+(* Gets a set of vars that get updated in the body of a loop. *)
+let loop_vars (sub : Sub.t) : Var.Set.t =
+  let visitor =
+    (object inherit [Var.Set.t] Term.visitor
+      method! visit_def def vars =
+        Var.Set.add vars (Def.lhs def)
+    end)
+  in
+  visitor#visit_sub sub Var.Set.empty
+
+(* Creates fresh names for [vars] found in the constraint. *)
+let freshen (constr : Constr.t) (env : Env.t) (vars : Var.Set.t)
+  : Constr.t * Env.t =
+  Var.Set.fold vars ~init:(constr, env) ~f:(fun (constr, env) v ->
+      let z3_v, env = Env.get_var env v in
+      let name = Format.sprintf "fresh_%s" (Expr.to_string z3_v) in
+      let fresh = Env.new_z3_expr ~name env (Var.typ v) in
+      Constr.substitute_one constr z3_v fresh, Env.add_call_pred env fresh)
+
+(* Gets the precondition of the exit node of a loop. *)
+let exit_pre (env : Env.t) (post : Constr.t) (node : Graphs.Ir.Node.t)
+    (g : Graphs.Ir.t) (loop : Graphs.Ir.Node.t group) : Constr.t =
+  let p =
+    List.find_map [Env.loop_exit] ~f:(fun spec ->
+        let* exit = spec node g in
+        let tid = exit |> Graphs.Ir.Node.label |> Term.tid in
+        let skip_node n =
+          let label = Graphs.Ir.Node.label n in
+          let in_loop = in_loop loop (Term.tid label) in
+          let unreachable = unreachable_from_start g exit n in
+          in_loop || unreachable
+        in
+        let _, env = visit_graph env post ~start:exit ~skip_node g in
+        Env.get_precondition env tid)
+  in
+  match p with
+  | Some p -> p
+  | None ->
+    let tid = node |> Graphs.Ir.Node.label |> Term.tid in
```

Comment:
Add a test where we would reach this point where we can't find the precondition.